File: Sources/HuChuanCore/Chunks.swift
import Foundation

public struct ChunkPlan: Equatable, Sendable {
    public let fileSize: UInt64
    public let chunkSize: UInt32

    public init(fileSize: UInt64, chunkSize: UInt32) {
        self.fileSize = fileSize
        self.chunkSize = max(chunkSize, 1)
    }

    public var chunkCount: Int {
        if fileSize == 0 { return 1 }
        return Int((fileSize + UInt64(chunkSize) - 1) / UInt64(chunkSize))
    }

    public func byteRange(for index: Int) -> Range<UInt64> {
        let start = UInt64(index) * UInt64(chunkSize)
        let end = min(start + UInt64(chunkSize), fileSize)
        return start..<end
    }

    public func length(for index: Int) -> Int {
        let r = byteRange(for: index)
        return Int(r.upperBound - r.lowerBound)
    }

    public static func plan(fileSize: UInt64) -> ChunkPlan {
        let size: UInt32
        if fileSize == 0 {
            size = 1
        } else if fileSize < 2_000_000 {
            size = UInt32(fileSize)
        } else if fileSize < 64_000_000 {
            size = 2 * 1024 * 1024
        } else if fileSize < 1_000_000_000 {
            size = 4 * 1024 * 1024
        } else {
            size = 8 * 1024 * 1024
        }
        return ChunkPlan(fileSize: fileSize, chunkSize: size)
    }

    public static func parallelConnections(fileSize: UInt64, configuredMax: Int) -> Int {
        let cap = max(1, min(configuredMax, 16))
        if fileSize < 8_000_000 { return 1 }
        if fileSize < 32_000_000 { return min(2, cap) }
        if fileSize < 128_000_000 { return min(4, cap) }
        return cap
    }
}

/// 紧凑位图：第 i 位为 1 表示该分块已收齐。
public struct ChunkBitmap: Equatable, Sendable {
    public private(set) var bytes: [UInt8]
    public let chunkCount: Int

    public init(chunkCount: Int) {
        self.chunkCount = max(chunkCount, 1)
        self.bytes = [UInt8](repeating: 0, count: (self.chunkCount + 7) / 8)
    }

    public init(chunkCount: Int, data: Data) {
        self.chunkCount = max(chunkCount, 1)
        let need = (self.chunkCount + 7) / 8
        var raw = [UInt8](data.prefix(need))
        if raw.count < need {
            raw.append(contentsOf: [UInt8](repeating: 0, count: need - raw.count))
        }
        self.bytes = raw
    }

    public var data: Data { Data(bytes) }

    public var base64: String { data.base64EncodedString() }

    public static func fromBase64(_ text: String, chunkCount: Int) -> ChunkBitmap {
        ChunkBitmap(chunkCount: chunkCount, data: Data(base64Encoded: text) ?? Data())
    }

    public func contains(_ index: Int) -> Bool {
        guard index >= 0, index < chunkCount else { return false }
        return bytes[index / 8] & (1 << (index % 8)) != 0
    }

    public mutating func insert(_ index: Int) {
        guard index >= 0, index < chunkCount else { return }
        bytes[index / 8] |= (1 << (index % 8))
    }

    public var receivedCount: Int {
        var n = 0
        for i in 0..<chunkCount where contains(i) { n += 1 }
        return n
    }

    public var isComplete: Bool { receivedCount == chunkCount }

    public func missingIndices() -> [Int] {
        (0..<chunkCount).filter { !contains($0) }
    }
}

public struct FileOffer: Codable, Equatable, Sendable, Identifiable {
    public var id: Int { index }
    public var index: Int
    public var relativePath: String
    public var size: UInt64
    public var modified: TimeInterval

    public init(index: Int, relativePath: String, size: UInt64, modified: TimeInterval) {
        self.index = index
        self.relativePath = relativePath
        self.size = size
        self.modified = modified
    }

    public var plan: ChunkPlan { ChunkPlan.plan(fileSize: size) }
}

File: win/protocol/chunks.go
package protocol

import "encoding/base64"

type ChunkPlan struct {
	FileSize  uint64
	ChunkSize uint32
}

func Plan(fileSize uint64) ChunkPlan {
	var size uint32
	switch {
	case fileSize == 0:
		size = 1
	case fileSize < 2_000_000:
		size = uint32(fileSize)
	case fileSize < 64_000_000:
		size = 2 * 1024 * 1024
	case fileSize < 1_000_000_000:
		size = 4 * 1024 * 1024
	default:
		size = 8 * 1024 * 1024
	}
	return ChunkPlan{FileSize: fileSize, ChunkSize: size}
}

func (p ChunkPlan) ChunkCount() int {
	if p.FileSize == 0 {
		return 1
	}
	cs := uint64(p.ChunkSize)
	if cs == 0 {
		cs = 1
	}
	return int((p.FileSize + cs - 1) / cs)
}

func (p ChunkPlan) ByteRange(index int) (start, end uint64) {
	cs := uint64(p.ChunkSize)
	if cs == 0 {
		cs = 1
	}
	start = uint64(index) * cs
	end = start + cs
	if end > p.FileSize {
		end = p.FileSize
	}
	return
}

func (p ChunkPlan) Length(index int) int {
	s, e := p.ByteRange(index)
	return int(e - s)
}

func ParallelConnections(fileSize uint64, configuredMax int) int {
	capn := configuredMax
	if capn < 1 {
		capn = 1
	}
	if capn > 16 {
		capn = 16
	}
	switch {
	case fileSize < 8_000_000:
		return 1
	case fileSize < 32_000_000:
		if capn < 2 {
			return capn
		}
		return 2
	case fileSize < 128_000_000:
		if capn < 4 {
			return capn
		}
		return 4
	default:
		return capn
	}
}

type ChunkBitmap struct {
	bytes      []byte
	chunkCount int
}

func NewBitmap(chunkCount int) ChunkBitmap {
	if chunkCount < 1 {
		chunkCount = 1
	}
	return ChunkBitmap{bytes: make([]byte, (chunkCount+7)/8), chunkCount: chunkCount}
}

func BitmapFromData(chunkCount int, data []byte) ChunkBitmap {
	b := NewBitmap(chunkCount)
	n := len(b.bytes)
	if len(data) < n {
		copy(b.bytes, data)
	} else {
		copy(b.bytes, data[:n])
	}
	return b
}

func BitmapFromBase64(text string, chunkCount int) ChunkBitmap {
	raw, err := base64.StdEncoding.DecodeString(text)
	if err != nil {
		return NewBitmap(chunkCount)
	}
	return BitmapFromData(chunkCount, raw)
}

func (b ChunkBitmap) Data() []byte { return append([]byte(nil), b.bytes...) }

func (b ChunkBitmap) Base64() string { return base64.StdEncoding.EncodeToString(b.bytes) }

func (b ChunkBitmap) Contains(index int) bool {
	if index < 0 || index >= b.chunkCount {
		return false
	}
	return b.bytes[index/8]&(1<<uint(index%8)) != 0
}

func (b *ChunkBitmap) Insert(index int) {
	if index < 0 || index >= b.chunkCount {
		return
	}
	b.bytes[index/8] |= 1 << uint(index%8)
}

func (b ChunkBitmap) ReceivedCount() int {
	n := 0
	for i := 0; i < b.chunkCount; i++ {
		if b.Contains(i) {
			n++
		}
	}
	return n
}

func (b ChunkBitmap) IsComplete() bool { return b.ReceivedCount() == b.chunkCount }
